File: 프로그래머스/2/92341. 주차 요금 계산/주차 요금 계산.py
from collections import defaultdict
from math import ceil


def convert_time(time):
    return int(time[:2]) * 60 + int(time[3:])


def calculate_fee(duration, fees):
    if duration <= fees[0]:
        return fees[1]

    return fees[1] + ceil((duration - fees[0]) / fees[2]) * fees[3]


def solution(fees, records):
    history = defaultdict(int)
    accumulate_time = defaultdict(int)
    total_fee = defaultdict(int)
    
    close_time = convert_time("23:59")

    for record in records:
        time, car_number, condition = record.split(' ')
        time = convert_time(time)

        if condition == 'IN':
            history[car_number] = time
            continue

        accumulate_time[car_number] += time - history[car_number]
        del history[car_number]
    
    for car_number in history:
        accumulate_time[car_number] += close_time - history[car_number]
        
    for car_number in accumulate_time:
        total_fee[car_number] = calculate_fee(accumulate_time[car_number], fees)

    return [value for _, value in sorted(total_fee.items())]

File: 프로그래머스/2/92341. 주차 요금 계산/주차 요금 계산.java
import java.util.stream.Collectors;
import java.util.HashMap;
import java.util.Map;

class Solution {
    
    public int[] solution(int[] fees, String[] records) {
        Map<String, Integer> accumulateTimes = calculateAccumulateTimes(records);
        Map<String, Integer> totalFees = calculateTotalFees(accumulateTimes, fees);
        
        return totalFees.entrySet()
                .stream()
                .sorted(Map.Entry.comparingByKey())
                .mapToInt(Map.Entry::getValue)
                .toArray();
    }
    
    private Map<String, Integer> calculateTotalFees(Map<String, Integer> accumulateTimes, int[] fees) {
        Map<String, Integer> totalFees = new HashMap<>();
        
        for (String carNumber : accumulateTimes.keySet()) {
            totalFees.put(
                    carNumber,
                    calculateFee(accumulateTimes.getOrDefault(carNumber, 0), fees));
        }
            
        return totalFees;
    }
    
    private int calculateFee(int duration, int[] fees) {
        if (duration <= fees[0]) {
            return fees[1];
        }
        
        return fees[1] + (int) Math.ceil((double) (duration - fees[0]) / fees[2]) * fees[3];
    }
    
    private Map<String, Integer> calculateAccumulateTimes(String[] records) {
        Map<String, Integer> history = new HashMap<>();
        Map<String, Integer> accumulateTimes = new HashMap<>();
        
        int closeTime = convertTime("23:59");
        
        for (String record : records) {
            String[] target = record.split(" ");
            
            int time = convertTime(target[0]);
            String carNumber = target[1];
            
            if (target[2].equals("IN")) {
                history.put(carNumber, time);
                continue;
            }
            
            accumulateTimes.put(
                    target[1], 
                    accumulateTimes.getOrDefault(carNumber, 0) + time - history.get(carNumber));
            
            history.remove(carNumber);
        }
        
        for (String carNumber : history.keySet()) {
            accumulateTimes.put(
                    carNumber, 
                    accumulateTimes.getOrDefault(carNumber, 0) + closeTime - history.get(carNumber));
        }
        
        return accumulateTimes;
    }
    
    private int convertTime(String time) {
        String hour = time.substring(0, 2);
        String minute = time.substring(3, 5);
        
        return Integer.parseInt(hour) * 60 + Integer.parseInt(minute);
    }
}